sklearn/preprocessing/imputation.py

```diff
@@ -366,13 +397,37 @@ def transform(self, X):
 
             mask = _get_mask(X, self.missing_values)
             n_missing = np.sum(mask, axis=self.axis)
-            values = np.repeat(valid_statistics, n_missing)
 
-            if self.axis == 0:
-                coordinates = np.where(mask.transpose())[::-1]
+            if self.strategy == 'knn':
+                if self.axis == 1:
+                    X = X.transpose()
+                    mask = mask.transpose()
+                    statistics = statistics.transpose()
+                missing_index = np.where(mask.any(1))[0]
+                for i, row in zip(missing_index, X[missing_index]):
+                    col_index = np.where(np.logical_not(np.isnan(row)))[0]
+                    impute_index = np.where(np.isnan(row))[0]
+                    neigh = NearestNeighbors(self.kneighbor)
```

Comment:
I just did a speed test by using make_classification function to create dataset. 40% of samples have 1 missing feature, 40% of samples have 2 missing features and the rest is full. KNN imputation used 6 neighbors.

10000 samples, 20 features:
-mean: 0.0036s
-@jnothman 's method: 5.6968s
-default KNN: 14.3168s

10000 samples, 200 features:
-mean: 0.0036
-@jnothman 's method: 
Didn't make it work... Used up all memory and disk space and then my mac stuck....
-default KNN: 68.7889s